Catalog Examples Document

reshmabidikar opened the pull request:
Catalog Examples document. Includes the following:

- Examples of catalogs with different billing modes 
- Examples of catalogs with different plan phases
- Examples of catalogs with multiple plans
- Examples of catalogs with different billing alignment rules

reshmabidikar committed:
Work in progress

reshmabidikar committed:
With examples for Billing modes, plan phases and account billing
alignment

reshmabidikar committed:
Catalog examples document with examples for subscription billing
alignment, bundle billing alignment

reshmabidikar committed:
Overall corrections in catalog examples document and sample catalogs

reshmabidikar committed:
Overall corrections

reshmabidikar committed:
Minor corrections

reshmabidikar committed:
Merge branch 'killbill:v3' into v3

reshmabidikar committed:
Deleted catalog with multiple billing alignment rules

FlowWriter reviewed (commented):
I didn't see anything amiss in these XML catalogs. I do think that eventually it would be helpful to either create a read-me file that lists the different types with a summarized explanation of each OR have a block of commented code in the beginning that summarizes its main points.

ghost commented:
Okay - disregard the previous reply. I obviously missed reviewing a large chunk of material...

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|===
+|Plan Name   |Product|Plan Description
+//----------------------
+|standard_monthly   |Standard   |Single EVERGREEN phase with a monthly recurring charge of $24.95   
+|standard_annual  |Standard   |Single EVERGREEN phase with an annual recurring charge of $275  
```
I would keep using the dash instead of the underscore as you have in the previous sections: standard-monthly, standard-annual

reshmabidikar replied:
Agreed, I think it was a typo

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Create a new account. Specify BCD as `25`  |Account created with BCD set to `25`
+|Create a subscription for the `standard-monthly` plan on `2021-09-17`   |Invoice generated for `$24.95` for `2021-09-17` to `2021-10-17` (BCD is ignored, subscription is billed as per its creation date)
+|Trigger invoice dry run with target date `2021-10-17`   |Invoice generated for `$24.95` for `2021-10-17` to `2021-11-17` 
+|Create a subscription for the `standard-annual` plan with date as `2021-09-30`   |No invoice generated
```
```suggestion
|Create a subscription for the `standard-annual` plan with date of `2021-09-30`   |No invoice generated
```

FlowWriter replied:
Why is the invoice not generated here?

reshmabidikar replied:
The invoice is not generated because a future date is specified, I've now made this clear in the document.

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Trigger invoice dry run with target date `2021-10-17`   |Invoice generated for `$24.95` for `2021-10-17` to `2021-11-17` 
+|Create a subscription for the `standard-annual` plan with date as `2021-09-30`   |No invoice generated
+|Trigger invoice dry run with target date `2021-09-30`   |Invoice generated for `$275` for `2021-09-30` to `2022-09-30` 
+|Trigger invoice dry run with target date `2021-10-17`   |Invoice generated for `$24.95` for `2021-10-17` to `2021-11-17` corresponding to `standard-monthly` plan 
```
```suggestion
|Trigger invoice dry run with target date `2021-11-17`   |Invoice generated for `$24.95` for `2021-11-17` to `2021-12-17` corresponding to `standard-monthly` plan 
```

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Trigger invoice dry run with target date `2022-09-30`   |Invoice generated for `$275` for `2022-09-30` to `2023-09-30` corresponding to `standard-annual` plan  
+|===
+
+As mentioned above, even if there is an addon plan associated with a subscription, the addon plan will be invoiced as per its own date and NOT be aligned with the subscription billing date.
```
```suggestion
As mentioned above, even if there is an add-on plan associated with a subscription, the add-on plan is invoiced as per its own date and is NOT aligned with the subscription billing date.
```

reshmabidikar replied:
I've removed this text and added a note at the top of this section.

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+When `BUNDLE` billing alignment is specified in the catalog, all subscriptions in a bundle are invoiced together. The bill day is derived from the first bill date of the first subscription. 
+
+This example uses the  https://github.com/killbill/killbill-docs/blob/1a945188c0ef53a3c7f797b4890514d35c09c46f/catalogs/bundle-billing-alignment.xml[Bundle Billing Alignment] catalog. It specifies `BUNDLE` billing alignment. 
```
```suggestion
This example uses the https://github.com/killbill/killbill-docs/blob/1a945188c0ef53a3c7f797b4890514d35c09c46f/catalogs/bundle-billing-alignment.xml[Bundle Billing Alignment] catalog. It specifies `BUNDLE` billing alignment. 
```

FlowWriter replied:
There is an extra space here. Please delete.

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Create a new account. Specify BCD as `25`  |Account created with BCD set to `25`
+|Create a subscription for the `standard-monthly` plan on `2021-09-20`   |Invoice generated for `$24.95` for `2021-09-20` to `2021-10-20` (BCD is ignored, subscription is billed as per its creation date)
+|Trigger invoice dry run with target date `2021-10-20`   |Invoice generated for `$24.95` for `2021-10-20` to `2021-11-20`
+|Add the `remotecontrol-monthly` add-on to the existing subscription with date as `2021-09-30`   |No invoice generated
```
```suggestion
|Add the `remotecontrol-monthly` add-on to the existing subscription with date of `2021-09-30`   |No invoice generated
```

FlowWriter replied:
So here the system does not auto-generate the invoice because it's an add-on product?

reshmabidikar replied:
@FlowWriter, no, the system does not auto-generate the invoice because it's a future date. I think we need to make this clear in the document. I'll make changes.

FlowWriter replied:
Did you clarify this somewhere?

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+//-------------
+|Create a new account. Leave BCD field empty  |Account created, BCD is blank.
+|Create a subscription on `2021-09-17`   |Invoice generated for `$24.95` for `2021-09-17` to `2021-10-17`. BCD set to `17` (Since subscription is created on 17th)
+|Create a subscription, with date as `2021-09-25`   |No invoice generated
```
So here the system doesn't generate the invoice because the account BCD is the 17th?

reshmabidikar replied:
@FlowWriter, the system does not generate an invoice since a future date is specified. I've now made this clear in the document.

FlowWriter replied:
Okay

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Create a subscription for the `standard-monthly` plan on `2021-09-20`   |Invoice generated for `$24.95` for `2021-09-20` to `2021-10-20` (BCD is ignored, subscription is billed as per its creation date)
+|Trigger invoice dry run with target date `2021-10-20`   |Invoice generated for `$24.95` for `2021-10-20` to `2021-11-20`
+|Add the `remotecontrol-monthly` add-on to the existing subscription with date as `2021-09-30`   |No invoice generated
+|Trigger invoice dry run with target date `2021-09-30`   |Invoice generated for `$11.97` for `2021-09-30` to `2021-10-20` corresponding to the `remotecontrol-monthly` plan (Prorated since `BUNDLE` billing alignment is used) 
+|Trigger invoice dry run with target date `2021-10-20`   |Invoice generated for `$42.90` for `2021-10-20` to `2021-11-20` corresponding to both plans
```
Is this a case of line 367 OR line 368? It's not clear to me if 367 is an optional scenario in which the business decides to use proration, whereas line 368 is a scenario where 367 (proration) is skipped and the business just waits until Oct 20th to bill.

reshmabidikar replied:
In the case of 367, the business uses proration since the addon needs to be billed along with the base plan (The base plan is billed on the 20th and `BUNDLE` billing alignment is used). After that, all subsequent invoices are billed on the 20th (which is what 368 demonstrates).

FlowWriter replied:
Got it!

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Trigger invoice dry run with target date `2021-10-20`   |Invoice generated for `$42.90` for `2021-10-20` to `2021-11-20` corresponding to both plans
+|===
+
+Note that `BUNDLE` billing alignment applies only to subscriptions that are part of the same bundle and does not affect subscriptions created in different bundles. So, when multiple subscriptions are created in different bundles, each one is still billed as per its creation date.
```
```suggestion
[NOTE]
*Note:* `BUNDLE` billing alignment applies only to subscriptions within a single bundle. This type of billing alignment does not affect subscriptions across different bundles. So, when multiple subscriptions are created in different bundles, each one is still billed as per its creation date.
```

FlowWriter replied:
I wanted to make this more clear; not sure if I did. I'm not sure what "one" in the last sentence refers to. It could refer to "bundle" or "subscription." 

Also, this note should go at the beginning of the "BUNDLE Billing Alignment" section because we can't assume readers will read to the very end of a section.

reshmabidikar replied:
@FlowWriter, "one" refers to subscriptions.  I had not added the text at the beginning as I did not want to confuse users with too much information at the start. However, the [note] feature is good, I wasn't aware of it, it differentiates the "note" from the main text. So, I've now added the note to the start of the document.

FlowWriter replied:
I think the last sentence (since it's repeating what's mentioned in the beginning of the section) isn't necessary. I still think this note belongs in the beginning of the section because you can't assume the reader will read all the way to the end of the section.

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+== Additional Information
+
+https://docs.killbill.io/latest/userguide_subscription.html#components-catalog[Subscription Billing User Guide]
```
```suggestion
https://docs.killbill.io/latest/userguide_subscription.html#components-catalog[_Subscription Billing User Guide_]
```

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+https://docs.killbill.io/latest/userguide_subscription.html#components-catalog[Subscription Billing User Guide]
+
+https://docs.killbill.io/latest/invoice_subsystem.html[Invoice Subsystem]
```
```suggestion
https://docs.killbill.io/latest/invoice_subsystem.html[_Invoice Subsystem_] manual
```

FlowWriter reviewed (commented):
@reshmabidikar - This is an amazing guide! It is going to be so helpful for KB users. 

I have made many suggested changes, most for the sake of consistency. Let me know if I need to provide add'l explanations. 

I'd like to review again before we release, but if Pierre/Stéphane are anxious to release it, then go ahead.

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+= Catalog Examples
+
+This document provides examples of catalogs and how they affect  invoices/subscriptions. 
```
Fixed

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+You can try out the various examples in each section by:
+
+* Uploading the corresponding catalog (via Kaui or https://killbill.github.io/slate/#catalog-upload-a-catalog-as-xml[Upload Catalog] API call)
+* Creating an account (via Kaui or https://killbill.github.io/slate/#account-create-an-account[Create Account] API call)
+
+* Creating a subscription as required (via Kaui or https://killbill.github.io/slate/#subscription-create-a-subscription[Create Subscription] API call)
+* Triggering an invoice dry-run (via Kaui or https://killbill.github.io/slate/#invoice-generate-a-dry-run-invoice[Generate Dry Run] API call)
```
Fixed

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+== Prerequisites
+
+Ensure that you have gone through the https://docs.killbill.io/latest/userguide_subscription.html#components-catalog[Subscription User Guide] and are familiar with the various catalog components like products, plans, phases.
```
Fixed

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+== Billing Modes
+
+A catalog needs to include a *Billing Mode* as explained in the https://docs.killbill.io/latest/invoice_subsystem.html#_overview[invoicing] document. 
```
Fixed

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+A catalog needs to include a *Billing Mode* as explained in the https://docs.killbill.io/latest/invoice_subsystem.html#_overview[invoicing] document. 
+
+Quick summary about Billing modes:
```
Fixed

reshmabidikar committed:
Changes after Mary's review

reshmabidikar commented:
> @reshmabidikar - This is an amazing guide! It is going to be so helpful for KB users. In fact, it was very helpful for me!
> 
> I have made many suggested changes, most for the sake of consistency. Let me know if I need to provide add'l explanations.
> 
> I'd like to review again before we release, but if Pierre/Stéphane are anxious to release it, then go ahead.

@FlowWriter, I've addressed most of your comments, please feel free to take another look.

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+* *IN_ARREAR* indicates that each recurring subscription is charged at the _end_ of the billing period.
+
+This section provides examples of catalogs with different billing modes. For simplicity sake, only catalogs with catalog level billing modes are demonstrated below, however the behavior is the same even when a per-plan billing mode is used.
```
@reshmabidikar - Regarding catalog-level versus plan-level billing modes, should we refer them to https://docs.killbill.io/latest/userguide_subscription.html#components-catalog-usage?

reshmabidikar replied:
@FlowWriter, I may be wrong, but I think per-plan billing is not the same as https://docs.killbill.io/latest/userguide_subscription.html#components-catalog-usage. Per-plan billing refers to specifying the billing mode (`IN_ARREAR` or `IN_ADVANCE`) at the plan level instead of the catalog level as explained [here](https://docs.killbill.io/latest/invoice_subsystem.html#_overview) whereas https://docs.killbill.io/latest/userguide_subscription.html#components-catalog-usage refers to usage billing. @pierre , can you please confirm if this is correct?

pierre replied:
Yup, correct!

_Historical fun fact: we introduced the ability to set the billing mode at the plan level when we first implemented usage-based pricing. Before, only a single global billing mode was supported._

FlowWriter commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+//-------------
+|Create a subscription on `2021-09-15`   |Invoice generated for `$4.95` due to `DISCOUNT` phase
+|Trigger invoice dry run with target date `2021-10-15`    |Invoice generated for  `$4.95` for `2021-10-15` to `2021-11-15`  
+|Trigger invoice dry run with target date `2021-12-15`    |Discount phase ends, Invoice generated for  `$24.95` for `2021-12-15` to `2022-01-15`   
```
Please change "Invoice" to "invoice."

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+
+This section provides examples of catalogs with different billing modes. For simplicity sake, only catalogs with catalog level billing modes are demonstrated below, however the behavior is the same even when a per-plan billing mode is used.
+
+Also, for simplicity sake, all the catalogs demonstrated in this section include a single product (`STANDARD`) and a single plan (`standard-monthly`) corresponding to it.
```
Fixed

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|===
+|Action   |Result   
+//-------------
+|Create a subscription on `2021-09-17`   |Invoice generated for `$24.95` for  `2021-09-17` to `2021-10-17` since *IN_ADVANCE* billing is used.
```
@FlowWriter, while this is a good suggestion, I will skip this for now as this would require making changes to the entire document.

FlowWriter replied:
Okay

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|Create a subscription on `2021-09-10`   |Invoice generated for `$24.95` for `2021-09-10` to `2021-10-10` 
+|Trigger invoice dry run with target date `2021-09-18`    |Invoice generated for  `$24.95` for `2021-09-17` to `2021-09-24`  
+|Trigger invoice dry run with target date `2021-10-15`    |Invoice generated for  `$24.95` for `2021-10-15` to `2021-10-22`   
+|Trigger invoice dry run with target date `2021-10-29`    |No invoice generated as the fixed term is over.  
```
@FlowWriter, agreed, specifying the time periods as `Sept 10 - Sept 16` or using a calendar would be better than just mentioning dates.  However, let's keep this on hold for now, my first priority would be to complete the document for all the different catalog components. I am currently working on round 2 of the document that explains the other catalog components (plan change rules, plan cancellation rules, and so on) and I'd like to complete that first. @pierre, I'd be interested in knowing what you think.

FlowWriter replied:
No worries!

pierre replied:
Yup, that makes sense to me: let's get the content right first, and then let's try to make it a bit more visually appealing.

reshmabidikar commented on catalogs/monthly-with-trial.xml:
```diff
@@ -31,7 +31,7 @@
-                        <number>30</number>
+                        <number>10</number>
                     </duration>
                     <fixed>
```
AFAIK,  it shouldn't matter. @pierre, could you please confirm if  `<billingPeriod>NO_BILLING_PERIOD</billingPeriod>`  is required?

pierre replied:
Yup, not explicitly needed here.

reshmabidikar commented on userguide/tutorials/catalog-examples.adoc:
```diff
@@ -0,0 +1,380 @@
+|===
+|Action   |Result   
+//-------------
+|Create a subscription on `2021-09-10`   |Invoice generated for `$0` due to trial phase
```
Not sure I understand this one :). Do you mean that it is possible to hide a $0 invoice in Kaui?

FlowWriter replied:
Actually, what I meant was that the Kill Bill customer could avoid displaying it to their customer. But I don't think this is the right document for this information, so you can close it out. :-)

reshmabidikar committed:
Changes as per Mary's comments

reshmabidikar commented:
Hi @FlowWriter, I had missed addressing some of your review comments yesterday. I've now addressed most of your comments.

ghost commented:
Okay, @reshmabidikar, I went through here again and made comments to help you close things out. I think we're just waiting on @pierre to answer a couple of questions.

pierre commented:
😍
